### PR TITLE
add secret finalizer resource

### PR DESCRIPTION
**Pull request by xh3b4sd**
Fixes https://github.com/giantswarm/giantswarm/issues/6057.

### CODE REVIEW

File: helm/aws-operator-chart/templates/01-rbac.yaml [hunk 1]
```diff
@@ -84,6 +84,7 @@ rules:
       - secrets
     verbs:
       - get
+      - update
```
Comment (xh3b4sd) [inline helm/aws-operator-chart/templates/01-rbac.yaml]: The operator needs permissions to update finalizers in secrets.

File: service/controller/clusterapi/v27/cluster_resource_set.go [hunk 2]
```diff
@@ -45,6 +45,7 @@ import (
 	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v27/resource/routetable"
 	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v27/resource/s3bucket"
 	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v27/resource/s3object"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v27/resource/secretfinalizer"
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v27/cluster_resource_set.go]: Wiring the new resource in the clusterapi controller.

File: service/controller/clusterapi/v27/cluster_resource_set.go [hunk 3]
```diff
@@ -599,6 +613,7 @@ func NewClusterResourceSet(config ClusterResourceSetConfig) (*controller.Resourc
 		namespaceResource,
 		serviceResource,
 		endpointsResource,
+		secretFinalizerResource,
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v27/cluster_resource_set.go]: It is run last to ensure the finalizer is cleaned up last.

File: service/controller/clusterapi/v27/resource/secretfinalizer/delete.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// When the operator's resource implementations request the CR's finalizers to
+	// be kept, their deletion logic is delayed. That implies that we should not
+	// remove the secret finalizers here already, since certain resource
+	// implementations may still require secrets to be available during their own
+	// deletion logic execution in upcoming reconciliation loops.
+	if finalizerskeptcontext.IsKept(ctx) {
```
Comment (xh3b4sd) [inline service/controller/clusterapi/v27/resource/secretfinalizer/delete.go]: I think this makes sense. We should actually do this in `app-operator` the same way I think. Question is only how the primitives change when we go with some reconcile result. /cc @kopiczko @tomahawk28

File: service/controller/legacy/v27/cluster_resource_set.go [hunk 5]
```diff
@@ -46,6 +46,7 @@ import (
 	"github.com/giantswarm/aws-operator/service/controller/legacy/v27/resource/routetable"
 	"github.com/giantswarm/aws-operator/service/controller/legacy/v27/resource/s3bucket"
 	"github.com/giantswarm/aws-operator/service/controller/legacy/v27/resource/s3object"
+	"github.com/giantswarm/aws-operator/service/controller/legacy/v27/resource/secretfinalizer"
```
Comment (xh3b4sd) [inline service/controller/legacy/v27/cluster_resource_set.go]: Doing the same for the legacy controller.

File: service/controller/clusterapi/v27/resource/secretfinalizer/delete.go (hunk 4 above)
Comment (kopiczko) [reply service/controller/clusterapi/v27/resource/secretfinalizer/delete.go]: That's clever. We need to make that doable in the future as well. This is useful.

**Pull request merged**